flutter/flutter-provider/state-management-in-flutter-using-provider-Part2.md

```diff
@@ -0,0 +1,209 @@
+
+### **FutureProvider** 
+
+`FutureProvider` gets the value of a future and provides to its dependents. We can use `FutureProvider<T>` to get the `Future<T>` value and this value represents the state.
+
+```dart
+FutureProvider<int>(
+  initialData: 0,
+  builder: (context) {
+    return Future.delayed(Duration(seconds: 4), () => 1);
+  },
+  child: ...
+),
+```
+This, `FutureProvider<int>` gives the binary value 0 and 1, in the initial state it provides the 0, and as soon as future resolves (after 4 seconds)this provides the 1 as value
+
+### How to use provider in our code
+
+Provider has some methods
+
+- `context.watch<T>()`, which makes the widget listen to changes on `T`
+- `context.read<T>()`, which returns `T` without listening to it
+- `context.select<T, R>(R cb(T value))`, which allows a widget to listen to only a small part of `T`.
+
+Or we can use the `static` method `Provider.of<T>(context)` or  `Provider.of<T>(context, listen : false)` which will behave similarly to `watch`/`read` respectively
+
+### How these methods work
+
+These methods will look up in the widget tree starting from the widget associated with the `BuildContext`, and will return the nearest variable of type `T` found (or throw if nothing is found).
+
+```dart
+class Home extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return Text(
+      context.watch<String>(),
+    );
+  }
+}
+```
+
+we must pass the `type<>` here in `watch()`, searches the same type of data in the widget tree. this `watch()` will listen to a state change, and rebuilds with the new value, whenever the state changes.
+
+
+Similarly, `read<T>()` will look for the variable of type `T` in the widget tree from it's `BuildContext` and retrieves the data.
+
+ In case of `select((T value))`, it only listen to those part of `T`, if state change happens in this part, only then it's dependents will update there value. It prevents widget to unnecessarily rebuilding if something other changes.
+
+ Now, one static method which works the same as the above method `Provider.of<T>(context)` this method performs the same as `watch()`, listen to changes in state `T` 
+
+ ```dart
+class Home extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+      String provider = Provider.of<String>(context);
+    return Text(provider
+    );
+  }
+}
+```
+
+`Provider.of<T>(context)` method has `listen`, which is set to `true` by default, which means if we are using this method it will listen to state change.
+
+If we set `listen` to `false` then, this will not listen to state change
+
+Alternatively, the same thing we can also obtain by using [Consumer](https://pub.dev/documentation/provider/latest/provider/Consumer-class.html) and [Selector](https://pub.dev/documentation/provider/latest/provider/Selector-class.html)
+
+### Few things about Consumer and Selector 
```

Comment:
Now I got confused. What is Consumer and Selector and why it is explained here? 
If this is concerned with the core concept of Provider, add it on "What we are going to learn section". If not, remove it and write a part 3 🥳